java/TypeQL.java

```diff
@@ -180,172 +193,257 @@ public static ThingVariable.Relation rel(String playerVar) {
     }
 
-    public static ThingConstraint.Value.Long lte(long value) {
-        return new ThingConstraint.Value.Long(LTE, value);
+    public static ThingConstraint.Predicate lt(UnboundConceptVariable variable) {
+        return new ThingConstraint.Predicate(new Predicate.ThingVariable(LT, variable.toThing()));
     }
 
-    public static ThingConstraint.Value.Double lte(double value) {
-        return new ThingConstraint.Value.Double(LTE, value);
+    public static ThingConstraint.Predicate lt(UnboundValueVariable variable) {
+        return new ThingConstraint.Predicate(new Predicate.ValueVariable(LT, variable.toValue()));
     }
 
-    public static ThingConstraint.Value.Boolean lte(boolean value) {
-        return new ThingConstraint.Value.Boolean(LTE, value);
+    public static ThingConstraint.Predicate lte(long value) {
+        return new ThingConstraint.Predicate(new Predicate.Long(LTE, value));
     }
 
-    public static ThingConstraint.Value.String lte(String value) {
-        return new ThingConstraint.Value.String(LTE, value);
+    public static ThingConstraint.Predicate lte(double value) {
+        return new ThingConstraint.Predicate(new Predicate.Double(LTE, value));
     }
 
-    public static ThingConstraint.Value.DateTime lte(LocalDateTime value) {
-        return new ThingConstraint.Value.DateTime(LTE, value);
+    public static ThingConstraint.Predicate lte(boolean value) {
+        return new ThingConstraint.Predicate(new Predicate.Boolean(LTE, value));
     }
 
-    public static ThingConstraint.Value.Variable lte(UnboundVariable variable) {
-        return new ThingConstraint.Value.Variable(LTE, variable);
+    public static ThingConstraint.Predicate lte(String value) {
+        return new ThingConstraint.Predicate(new Predicate.String(LTE, value));
     }
 
-    public static ThingConstraint.Value.String contains(String value) {
-        return new ThingConstraint.Value.String(CONTAINS, value);
+    public static ThingConstraint.Predicate lte(LocalDateTime value) {
+        return new ThingConstraint.Predicate(new Predicate.DateTime(LTE, value));
     }
 
-    public static ThingConstraint.Value.String like(String value) {
-        return new ThingConstraint.Value.String(LIKE, value);
+    public static ThingConstraint.Predicate lte(UnboundConceptVariable variable) {
+        return new ThingConstraint.Predicate(new Predicate.ThingVariable(LTE, variable.toThing()));
     }
 
+    public static ThingConstraint.Predicate lte(UnboundValueVariable variable) {
+        return new ThingConstraint.Predicate(new Predicate.ValueVariable(LTE, variable.toValue()));
+    }
+
+    public static ThingConstraint.Predicate contains(String value) {
+        return new ThingConstraint.Predicate(new Predicate.String(CONTAINS, value));
+    }
+
+    public static ThingConstraint.Predicate like(String value) {
+        return new ThingConstraint.Predicate(new Predicate.String(LIKE, value));
+    }
```

Comment:
should we make an `assign(Expression)` helper on here?

Reply:
Could do
But assigns in such a context are actually equality predicates

Reply:
I see what you mean, But how would you use it?
I can only think of :`valueVar("x").assign(assign(Expression.plus(...)))`
The introduction of the static `assign` doesn't remove the methods returning an Expression.

Reply:
It will be tricky to be perfectly congruent to existing code because everything we have so far is chainable ((?)because it's fully left-associative (?)). Arithmetic isn't.